src/NuGet.Protocol.Core.v3/ServiceIndexResourceV3Provider.cs

```diff
@@ -56,29 +59,61 @@ public ServiceIndexResourceV3Provider()
-                        _cache.AddOrUpdate(url, entry, (key, value) => entry);
+                        var messageHandlerResource = await source.GetResourceAsync<HttpHandlerResource>(token);
+                        if (credentials != null)
+                        {
+                            messageHandlerResource.ClientHandler.Credentials = credentials;
+                        }
 
-                        return new Tuple<bool, INuGetResource>(false, null);
-                    }
-                    catch (HttpRequestException)
-                    {
-                        var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
-                        _cache.AddOrUpdate(url, entry, (key, value) => entry);
+                        try
+                        {
+                            using (var client = new DataClient(messageHandlerResource))
+                            {
+                                json = await client.GetJObjectAsync(uri, token);
+                                if (HttpHandlerResourceV3.CredentialsSuccessfullyUsed != null && credentials != null)
+                                {
+                                    HttpHandlerResourceV3.CredentialsSuccessfullyUsed(uri, credentials);
+                                }
+
+                                break;
+                            }
+                        }
+                        catch (JsonReaderException)
+                        {
+                            var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
+                            _cache.AddOrUpdate(url, entry, (key, value) => entry);
 
-                        return new Tuple<bool, INuGetResource>(false, null);
+                            return new Tuple<bool, INuGetResource>(false, null);
+                        }
+                        catch (HttpRequestException)
+                        {
+                            var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
+                            _cache.AddOrUpdate(url, entry, (key, value) => entry);
+
+                            return new Tuple<bool, INuGetResource>(false, null);
+                        }
+                        catch (UnauthorizedAccessException)
+                        {
+                            credentials = null;
+                            if (HttpHandlerResourceV3.PromptForCredentials != null)
+                            {
+                                credentials = HttpHandlerResourceV3.PromptForCredentials(uri);
+                            }
+
+                            if (credentials == null)
+                            {
+                                var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
+                                _cache.AddOrUpdate(url, entry, (key, value) => entry);
+
+                                return new Tuple<bool, INuGetResource>(false, null);
+                            }
+                        }
```

Comment:
The change here is to catch UnauthorizedAccessException, prompt user for new credentials, then retry the request with the new credentials.